ooni/agent/scheduler.py

```diff
@@ -0,0 +1,409 @@
+        try:
+            yield task.run()
+        except DidNotRun:
+            log.debug("Did not run {0}".format(task.identifier))
+        except Exception as exc:
+            log.err("Failed to run task {0}".format(task.identifier))
+            log.exception(exc)
+
+
+class SchedulerService(service.MultiService):
+    """
+    This service is responsible for running the periodic tasks.
+    """
+    def __init__(self, director, interval=30):
+        service.MultiService.__init__(self)
+        self.director = director
+        self.interval = interval
+        self._looping_call = LoopingCall(self._should_run)
+        self._scheduled_tasks = []
+
+    def schedule(self, task):
+        self._scheduled_tasks.append(task)
+
+    def unschedule(self, task):
+        # We first cancel the task so the run lock is deleted
+        task.cancel()
+        self._scheduled_tasks.remove(task)
+
+    def refresh_deck_list(self):
+
+        to_enable = []
+        for deck_id, deck in deck_store.list_enabled():
+            if deck.schedule is None:
+                continue
+            to_enable.append((deck_id, deck.schedule))
+
+        # If we are not initialized we should not enable anything
+        if not config.is_initialized():
+            log.msg("We are not initialized skipping setup of decks")
+            to_enable = []
+
+        for scheduled_task in self._scheduled_tasks[:]:
+            if not isinstance(scheduled_task, RunDeck):
+                continue
+
+            info = (scheduled_task.deck_id, scheduled_task.schedule)
+            if info in to_enable:
+                # If the task is already scheduled there is no need to
+                # enable it.
+                log.msg("The deck {0} is already scheduled".format(deck_id))
+                to_enable.remove(info)
+            else:
+                # If one of the tasks that is scheduled is no longer in the
+                # scheduled tasks. We should disable it.
+                log.msg("The deck task {0} should be disabled".format(deck_id))
+                self.unschedule(scheduled_task)
+
+        for deck_id, schedule in to_enable:
+            log.msg("Scheduling to run {0}".format(deck_id))
+            self.schedule(RunDeck(self.director, deck_id, schedule))
+
+    def _task_did_not_run(self, failure, task):
+        failure.trap(DidNotRun)
+        log.debug("Did not run {0}".format(task.identifier))
+
+    def _task_failed(self, failure, task):
+        log.err("Failed to run {0}".format(task.identifier))
+        log.exception(failure)
+
+    def _task_success(self, result, task):
+        log.debug("Ran {0}".format(task.identifier))
+
+    def _should_run(self):
+        """
+        This function is called every self.interval seconds to check
+        which periodic tasks should be run.
+        """
+        for task in self._scheduled_tasks:
+            log.debug("Running task {0}".format(task.identifier))
+            d = task.run()
```

Comment:
Exactly, the locking mechanism is in place to ensure that it will lock when somebody has already triggered a run.

There is a tradeoff that should probably be noted somewhere, that this backlog of tasks waiting on the lock can be quite large for a task that requires a long time to complete running.

For example if a task requires 1 hour to complete you will have a backlog of 60*60/30= 120 run operations.